README.md
<a href="https://opensource.newrelic.com/oss-category/#community-project"><picture><source media="(prefers-color-scheme: dark)" srcset="https://github.com/newrelic/opensource-website/raw/main/src/images/categories/dark/Community_Project.png"><source media="(prefers-color-scheme: light)" srcset="https://github.com/newrelic/opensource-website/raw/main/src/images/categories/Community_Project.png"><img alt="New Relic Open Source community project banner." src="https://github.com/newrelic/opensource-website/raw/main/src/images/categories/Community_Project.png"></picture></a>

# New Relic Pixie Integration

The Pixie integration connects to the Pixie API and enables the New Relic plugin in Pixie. The plugin then sends data generated by a set of PxL scripts to New Relic using the OpenTelemetry Line Protocol. The integration takes the preset script of the New Relic plugin and enables them for the provided cluster. The integration can be configured to register a set of custom PxL scripts on top of the preset scripts.

Once the New Relic plugin is successfully enabled and the PxL scripts are registered, the integration exits with error code 0.

## Getting Started

Make sure you have a Pixie account and a New Relic account set up, and have collected the following information:

* [New Relic license key](https://docs.newrelic.com/docs/accounts/accounts-billing/account-setup/new-relic-license-key/)
* [Pixie API key](https://docs.pixielabs.ai/using-pixie/api-quick-start/#get-an-api-token) - Using the px CLI: `px api-key create -s`
* [Pixie Cluster ID](https://docs.pixielabs.ai/using-pixie/api-quick-start/#get-a-cluster-id) - Using the px CLI: `px get cluster --id`
* Pixie endpoint - For a [self-hosted Pixie](https://docs.px.dev/installing-pixie/install-guides/self-hosted-pixie/) setup, use `dev.withpixie.dev:443`
* Cluster name - Using the px CLI: `px get viziers`

## Building

```make build-container```

Docker is required to build the Pixie integration container image.

## Usage

```docker run --env-file ./.env -it newrelic/newrelic-pixie-integration:latest```

Define the following environment variables in the `env.list` file:

```
NR_LICENSE_KEY=
PIXIE_API_KEY=
PIXIE_CLUSTER_ID=
PIXIE_ENDPOINT=
CLUSTER_NAME=
```

The following environment variables are optional.

```
HTTP_SPAN_LIMIT=5000
DB_SPAN_LIMIT=1000
COLLECT_INTERVAL_SEC=10
EXCLUDE_PODS_REGEX=
EXCLUDE_NAMESPACES_REGEX=
SCRIPT_DIR=/scripts
VERBOSE=true
```

The `*_LIMIT` environment variable can be used to control the amount of data that is sent to New Relic. The `COLLECT_INTERVAL_SEC` environment variable sets the collection interval for any PxL script which doesn't already have a defaultFrequency set. Setting the interval to `-1` will disable sending that data to New Relic. The smallest valid interval is 2 seconds.

The `EXCLUDE_PODS_REGEX` and `EXCLUDE_NAMESPACES_REGEX` environment variables can be configured with [RE2 regular expressions](https://github.com/google/re2/wiki/Syntax) to not send observability data to New Relic for the matching pods and namespaces. When `EXCLUDE_NAMESPACES_REGEX` is provided, no data for the matching namespaces will be sent. When `EXCLUDE_PODS_REGEX` is provided, no data for the matching pods (independent of the namespace they are in) will be sent.

Note: If the `docker run` command fails, try disabling the all plugins in the Pixie admin UI (/admin/plugins) before re-running the command.

## Custom scripts

## Testing

### Unit tests

Run the unit tests locally:

```make test```

### End-to-end tests

After executing the command above Pixie data should be flowing into your New Relic account. Use the following NRQL queries to verify this:

**Metrics**

```
SELECT * FROM Metric WHERE instrumentation.provider='pixie'
```

**Spans**

```
SELECT * FROM Span WHERE instrumentation.provider='pixie'
```

## Custom scripts

The integration picks up custom scripts in the `SCRIPT_DIR` (set to `/scripts` by default). The custom scripts should be defined in a yaml file with the `.yaml` extension. The following fields are required:

* name (string): the name of the script
* description (string): description of the script
* frequencyS (int): frequency to execute the script in seconds
* scripts (string): the actual PxL script to execute
* addExcludes (optional boolean, `false` by default): add pod and namespace excludes to the custom script

[This tutorial](https://docs.pixielabs.ai/tutorials/integrations/otel/#write-the-pxl-script) explains how to write custom PxL scripts. Example of a custom script, eg. `/scripts/custom1.yaml`:

```
name: "custom1"
description: "Custom script 1"
frequencyS: 60
script: |
    import px

    df = px.DataFrame(table='http_events', start_time=px.plugin.start_time, end_time=px.plugin.end_time)

    ns_prefix = df.ctx['namespace'] + '/'
    df.container = df.ctx['container_name']
    df.pod = px.strip_prefix(ns_prefix, df.ctx['pod'])
    df.service = px.strip_prefix(ns_prefix, df.ctx['service'])
    df.namespace = df.ctx['namespace']

    df.status_code = df.resp_status

    df = df.groupby(['status_code', 'pod', 'container','service', 'namespace']).agg(
        latency_min=('latency', px.min),
        latency_max=('latency', px.max),
        latency_sum=('latency', px.sum),
        latency_count=('latency', px.count),
        time_=('time_', px.max),
    )

    df.latency_min = df.latency_min / 1000000
    df.latency_max = df.latency_max / 1000000
    df.latency_sum = df.latency_sum / 1000000

    df.cluster_name = px.vizier_name()
    df.cluster_id = px.vizier_id()
    df.pixie = 'pixie'

    px.export(
    df, px.otel.Data(
        resource={
        'service.name': df.service,
        'k8s.container.name': df.container,
        'service.instance.id': df.pod,
        'k8s.pod.name': df.pod,
        'k8s.namespace.name': df.namespace,
        'px.cluster.id': df.cluster_id,
        'k8s.cluster.name': df.cluster_name,
        'instrumentation.provider': df.pixie,
        },
        data=[
        px.otel.metric.Summary(
            name='http.server.duration',
            description='measures the duration of the inbound HTTP request',
            # Unit is not supported yet
            # unit='ms',
            count=df.latency_count,
            sum=df.latency_sum,
            quantile_values={
            0.0: df.latency_min,
            1.0: df.latency_max,
            },
            attributes={
            'http.status_code': df.status_code,
            },
        )],
    ),
    )
```

The `addExcludes` adds a filtering section to the script right before the call to `px.export`. For example if `EXCLUDE_PODS_REGEX` is set to `team-1-.*`, the following line will be added to the PxL script:

```
df = df[not px.regex_match('team-1-.*', df.pod)]
```

 The filtering code requires your PxL script to:

* Use `df` as the exported dataframe
* The dataframe `df` to have a `pod` field (for `EXCLUDE_PODS_REGEX`)
* The dataframe `df` to have a `namespace` field (for `EXCLUDE_NAMESPACE_REGEX`)

If the above requirements do not hold for your script, setting `addExcludes` to `true` will break the script.

## Script registration behaviour

The integration will consider scripts that start with `nri-` as managed by the integration. Scripts are registered per cluster and follow the `nri-<script name>-<cluster name>` pattern. The integration updates the scripts to bring them in-sync with the provided configuration. Scripts that are no longer present in the configuration are deleted. If you already have scripts that start with `nri-`, the integration will remove these if they are not specified in the integration configuration.

A cluster name can only be used for a single cluster. Scripts that follow the `nri-<script name>-<cluster name>` pattern with the same `<cluster-name>` but that were registered from another cluster will be overwritten or deleted.

## Support

New Relic hosts and moderates an online forum where customers can interact with New Relic employees as well as other users to get help and share best practices. Like all official New Relic open source projects, there's a related Community topic in the New Relic Explorers Hub. You can find this project's topic/threads here: https://discuss.newrelic.com/t/new-relic-pixie-integration/143646

## Contribute

We encourage your contributions to improve the Pixie integration! Keep in mind that when you submit your pull request, you'll need to sign the CLA via the click-through using CLA-Assistant. You only have to sign the CLA once per project.

If you have any questions, or to execute our corporate CLA (which is required if your contribution is on behalf of a company), drop us an email at opensource@newrelic.com.

**A note about vulnerabilities**

As noted in our [security policy](../../security/policy), New Relic is committed to the privacy and security of our customers and their data. We believe that providing coordinated disclosure by security researchers and engaging with the security community are important means to achieve our security goals.

If you believe you have found a security vulnerability in this project or any of New Relic's products or websites, we welcome and greatly appreciate you reporting it to New Relic through [HackerOne](https://hackerone.com/newrelic).

If you would like to contribute to this project, review [these guidelines](./CONTRIBUTING.md).

To all contributors, we thank you!  Without your contribution, this project would not be what it is today.

## License

The New Relic Pixie integration is licensed under the [Apache 2.0](http://apache.org/licenses/LICENSE-2.0.txt) License.

The integration also uses source code from third-party libraries. You can find full details on which libraries are used and the terms under which they are licensed in the third-party notices document.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 2.1.4

## What's Changed
- Update CHANGELOG.md by @juanjjaramillo in https://github.com/newrelic/newrelic-pixie-integration/pull/117
- build(deps): bump github.com/stretchr/testify from 1.8.3 to 1.8.4 by @dependabot in https://github.com/newrelic/newrelic-pixie-integration/pull/119
- build(deps): bump aquasecurity/trivy-action from 0.10.0 to 0.11.0 by @dependabot in https://github.com/newrelic/newrelic-pixie-integration/pull/121
- build(deps): bump github.com/sirupsen/logrus from 1.9.2 to 1.9.3 by @dependabot in https://github.com/newrelic/newrelic-pixie-integration/pull/120
- build(deps): bump aquasecurity/trivy-action from 0.11.0 to 0.11.2 by @dependabot in https://github.com/newrelic/newrelic-pixie-integration/pull/122
- build(deps): bump alpine from 3.18.0 to 3.18.2 by @dependabot in https://github.com/newrelic/newrelic-pixie-integration/pull/123
- build(deps): bump google.golang.org/grpc from 1.55.0 to 1.56.0 by @dependabot in https://github.com/newrelic/newrelic-pixie-integration/pull/124
- build(deps): bump github.com/lestrrat-go/jwx from 1.2.4 to 1.2.26 by @dependabot in https://github.com/newrelic/newrelic-pixie-integration/pull/125


**Full Changelog**: https://github.com/newrelic/newrelic-pixie-integration/compare/v2.1.3...v2.1.4

## 2.1.3 (2023-05-31)

- Update base Alpine image from 3.17.2 to 3.18
- Update dependencies
- Minor improvements

## 2.1.2 (2023-05-05)

- Update base Alpine image from 3.13 to 3.17.2
- Update pxapi from 0.3.1 to 0.4.1
- Update dependencies

## 2.0.4 (2023-03-16)

- Update Golang from v1.16 to v1.20.
- Update dependencies.

## 0.1.5 (2021-05-14)

- Perform query normalization on the MySQL and Postgresql data.

## 0.1.4 (2021-05-10)

- Properly detect US license keys.

## 0.1.3 (2021-05-06)

- Add user agent to OpenTelemetry requests.

## 0.1.2 (2021-05-05)

- Fix detected issues in v0.1.1

## 0.1.1 (2021-05-04)

- Small fixes & README updates

## 0.1.0 (2021-04-20)

- Initial version of the integration.


http_metrics.go
package main

import (
	"time"

	"px.dev/pxapi/types"
)

const httpMetricsPXL = `
import px

ns_per_ms = 1000 * 1000
ns_per_s = 1000 * ns_per_ms
window_ns = px.DurationNanos(10 * ns_per_s)



df = px.DataFrame(table='http_events', start_time='-10s')

df.timestamp = px.bin(df.time_, window_ns)

df.container = df.ctx['container_name']
df.pod = df.ctx['pod']
df.service = df.ctx['service']
df.namespace = df.ctx['namespace']

df.status_code = df.resp_status

df = df.groupby(['timestamp', 'status_code', 'pod', 'container','service', 'namespace']).agg(
    latency_min=('latency', px.min),
    latency_max=('latency', px.max),
    latency_sum=('latency', px.sum),
    latency_count=('latency', px.count)
)

px.display(df, 'http')
`

type HttpMetricData struct {
	Timestamp   time.Time
	Service     string
	Pod         string
	ClusterName string
	Namespace   string
	Container   string
	StatusCode  int64
	Min         float64
	Max         float64
	Sum         float64
	Count       float64
}

func HttpMetricsHandler(r *types.Record, t *TelemetrySender) error {
	namespace, service, pod := takeNamespaceServiceAndPod(r)
	sendHttpMetric(&HttpMetricData{
		Timestamp:   r.GetDatum("timestamp").(*types.Time64NSValue).Value(),
		Service:     service,
		Pod:         pod,
		Container:   r.GetDatum(colContainer).String(),
		ClusterName: t.ClusterName,
		Namespace:   namespace,
		StatusCode:  r.GetDatum("status_code").(*types.Int64Value).Value(),
		Min:         float64(r.GetDatum("latency_min").(*types.Int64Value).Value()) / 1000000,
		Max:         float64(r.GetDatum("latency_max").(*types.Int64Value).Value()) / 1000000,
		Sum:         float64(r.GetDatum("latency_sum").(*types.Int64Value).Value()) / 1000000,
		Count:       float64(r.GetDatum("latency_count").(*types.Int64Value).Value()),
	}, t)

	return nil
}

func sendHttpMetric(data *HttpMetricData, t *TelemetrySender) {
	_ = t.ExportHttpMetric(data)
}
